### PR TITLE
SDIT-1623 Restructure csip report from offender booking

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CSIPReport.kt [hunk 1]
```diff
@@ -30,14 +30,18 @@ data class CSIPReport(
   @GeneratedValue(generator = "CSIP_ID")
   val id: Long = 0,
 
+  @Column(name = "CSIP_SEQ")
+  val sequence: String? = null,
+
   @ManyToOne(fetch = FetchType.LAZY)
   @JoinColumn(name = "OFFENDER_BOOK_ID")
-  val offenderBooking: OffenderBooking?,
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CSIPReport.kt:35]: The db defines the offenderBooking as nullable - however, none of the entries in this table have a missing offenderBooking number.  Also, the trigger on the insert gets the prisonId from the offenderbooking and fails if this is null.
I am therefore not convinced it should be nullable....so have change this to non-nullable (if that makes sense)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CSIPReport.kt [hunk 2]
```diff
@@ -30,14 +30,18 @@ data class CSIPReport(
...
+  val offenderBooking: OffenderBooking,
+
+  @Column(name = "ROOT_OFFENDER_ID", nullable = false)
+  val rootOffenderId: Long,
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CSIPReport.kt:41]: Wasn't sure how to map ROOT_OFFENDER_ID - we're not interested in this value as we can get it from the offenderBooking, but it obviously needs mapping and setting.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CSIPReport.kt [hunk 3]
```diff
@@ -62,34 +61,35 @@ class CSIPReportBuilder(
...
     offenderBooking: OffenderBooking,
     type: String,
     location: String,
     areaOfWork: String,
+    reportedBy: Staff,
   ): CSIPReport =
     CSIPReport(
       offenderBooking = offenderBooking,
-      offender = offender,
+      rootOffenderId = offenderBooking.offender.rootOffenderId ?: offenderBooking.offender.id,
```
[the remark this answers is not in the document]
Comment (steverendell) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CSIPReport.kt:72]: maybe the root offender's own rootOffender field is null

**Review by steverendell — APPROVED** [review]
lgtm

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CSIPReport.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (karenmillermoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CSIPReport.kt:72]: will check that - cheers

**Pull request merged**